runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -2925,16 +2925,17 @@ J9::CodeGenerator::compressedReferenceRematerialization()
    TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
 
    static bool disableRematforCP = feGetEnv("TR_DisableWrtBarOpt") != NULL;
-
-   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
-      {
-      // We need this restriction because the compressedrefs remat opt
-      // removes decompression/compression sequences from loads/stores where there doesn't exist
-      // a gc point between the load and the store, and the load doesn't need to be dereferenced.
-      // In Guarded Storage, we can't not do a guarded load because the object that is loaded may
-      // not be in the root set, and as a consequence, may get moved.
+   bool generateReadBarrier = self()->comp()->getOption(TR_EnableFieldWatch);
+
+   // The compressedrefs remat opt removes decompression/compression sequences from
+   // loads/stores where there doesn't exist a gc point between the load and the store,
+   // and the load doesn't need to be dereferenced.
+   // The opt needs to be disabled for the following cases:
+   // 1. In Guarded Storage, we can't not do a guarded load because the object that is loaded may
+   // not be in the root set, and as a consequence, may get moved.
+   // 2. For read barriers, the vmhelpers are GC points and therefore the object might be moved
+   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads() || generateReadBarrier)
```

Comment:
why isn't the first part of this || part of the bool added above? it would seem you would have to repeat this idiom no? Doesn't this warrant a trace statement of some kind to say we disabling because of this?